src/loader.rs

```diff
@@ -142,76 +209,93 @@ impl std::fmt::Display for LoadError {
     }
 }
 
-impl YamlLoader {
-    fn insert_new_node(&mut self, node: (Yaml, usize)) {
-        // valid anchor id starts from 1
-        if node.1 > 0 {
-            self.anchor_map.insert(node.1, node.0.clone());
-        }
-        if self.doc_stack.is_empty() {
-            self.doc_stack.push(node);
-        } else {
-            let parent = self.doc_stack.last_mut().unwrap();
-            match *parent {
-                (Yaml::Array(ref mut v), _) => v.push(node.0),
-                (Yaml::Hash(ref mut h), _) => {
-                    let cur_key = self.key_stack.last_mut().unwrap();
-                    // current node is a key
-                    if cur_key.is_badvalue() {
-                        *cur_key = node.0;
-                    // current node is a value
-                    } else {
-                        let mut newkey = Yaml::BadValue;
-                        std::mem::swap(&mut newkey, cur_key);
-                        h.insert(newkey, node.0);
-                    }
-                }
-                _ => unreachable!(),
-            }
-        }
-    }
-
-    /// Load the given string as a set of YAML documents.
+/// A trait providing methods used by the [`YamlLoader`].
+///
+/// This trait must be implemented on YAML node types (i.e.: [`Yaml`] and annotated YAML nodes). It
+/// provides the necessary methods for [`YamlLoader`] to load data into the node.
+pub trait LoadableYamlNode: Clone + std::hash::Hash + Eq {
+    /// Create an instance of `Self` from a [`Yaml`].
     ///
-    /// The `source` is interpreted as YAML documents and is parsed. Parsing succeeds if and only
-    /// if all documents are parsed successfully. An error in a latter document prevents the former
-    /// from being returned.
-    /// # Errors
-    /// Returns `ScanError` when loading fails.
-    pub fn load_from_str(source: &str) -> Result<Vec<Yaml>, ScanError> {
-        Self::load_from_iter(source.chars())
-    }
+    /// Nodes must implement this to be built. The optional metadata that they contain will be
+    /// later provided by the loader and can be default initialized. The [`Yaml`] object passed as
+    /// parameter may be of the [`Array`] or [`Hash`] variants. In this event, the inner container
+    /// will always be empty. There is no need to traverse all elements to convert them from
+    /// [`Yaml`] to `Self`.
+    ///
+    /// [`Array`]: `Yaml::Array`
+    /// [`Hash`]: `Yaml::Hash`
+    fn from_bare_yaml(yaml: Yaml) -> Self;
```

Comment:
I don't know if it's worth it, but interestingly with the parametrized Yaml representation, you can define a type that is exactly "a shallow YAML tree of 1 level that doesn't contain any more content", which is `YamlData<()>`, so that `from_bare_yaml` could return a `YamlData<()>`, which would better reflect what it actually represents

Reply:
I don't really think that is worth it, as it would require more constraints on the `Node` generic at the time of loading. We would need to add `Node: From<YamlData<()>>`, which I don't need that we want.